platform/entanglement/src/services/LinkService.js

```diff
@@ -45,36 +45,24 @@ define(
             if (parentCandidate.getId() === object.getId()) {
                 return false;
             }
-            if (parentCandidate.getModel().composition.indexOf(object.getId()) !== -1) {
+            if ((parentCandidate.getModel().composition || []).indexOf(object.getId()) !== -1) {
                 return false;
             }
             return this.policyService.allow(
                 "composition",
                 parentCandidate.getCapability('type'),
                 object.getCapability('type')
-            );
+            ) && parentCandidate.hasCapability('composition');
         };
 
         LinkService.prototype.perform = function (object, parentObject) {
-            function findChild(children) {
-                var i;
-                for (i = 0; i < children.length; i += 1) {
-                    if (children[i].getId() === object.getId()) {
-                        return children[i];
-                    }
-                }
-            }
+            // Note that this was checked-for explicitly during validate step
```

Comment:
If validate must be called before perform, perhaps perform should just call validate?

Otherwise, the JSDoc for perform would be the ideal place to include such a comment-- someone who doesn't know the service shouldn't be expected the read the inner workings of the method to know how to use it.

Erm, probably my fault for not writing JSDoc in the first place, but how did this make it through review w/o JSDoc?

Pete

On Sep 22, 2015, at 1:47 PM, Victor Woeltjen <notifications@github.com<mailto:notifications@github.com>> wrote:


In platform/entanglement/src/services/LinkService.js<https://github.com/nasa/openmctweb/pull/98#discussion_r40141789>:

>          };
>
>          LinkService.prototype.perform = function (object, parentObject) {
> -            function findChild(children) {
> -                var i;
> -                for (i = 0; i < children.length; i += 1) {
> -                    if (children[i].getId() === object.getId()) {
> -                        return children[i];
> -                    }
> -                }
> -            }
> +            // Note that this was checked-for explicitly during validate step


I think this is useful because it may not be immediately obvious that validate may/must be invoked before perform; will toy with phrasing a little to clarify that. (In general, it's risky to expose methods that need to be called in some particular sequence for correctness, so at bare minimum I like to have some kind of breadcrumb in case that goes wrong.)

—
Reply to this email directly or view it on GitHub<https://github.com/nasa/openmctweb/pull/98/files#r40141789>.